packages/eslint-plugin-query/src/rules/exhaustive-deps.test.ts

```diff
@@ -0,0 +1,179 @@
+      `,
+      output: `
+        const a = 1;
+        const b = 2;
+        useQuery({ queryKey: [\`entity/\${a}\`, b], queryFn: () => api.getEntity(a, b) });
+      `,
+      errors: [{ messageId: 'missingDeps', data: { deps: 'b' } }],
+    },
+    {
+      name: 'should fail when dep exists inside setter and missing in queryKey',
+      code: `
+        const [id] = React.useState(1);
+        useQuery({
+            queryKey: ["entity"],
+            queryFn: () => {
+                const { data } = axios.get(\`.../\${id}\`);
+                return data;
+            }
+        });
+      `,
+      output: `
+        const [id] = React.useState(1);
+        useQuery({
+            queryKey: ["entity", id],
+            queryFn: () => {
+                const { data } = axios.get(\`.../\${id}\`);
+                return data;
+            }
+        });
+      `,
+      errors: [{ messageId: 'missingDeps', data: { deps: 'id' } }],
+    },
+    {
+      name: 'should fail when dep does not exist while having a complex queryKey',
+      code: `
+        const todoQueries = {
+          key: (a, b, c, d, e) => ({
+            queryKey: ["entity", a, [b], { c }, 1, true],
+            queryFn: () => api.getEntity(a, b, c, d, e)
+          })
+        }
+      `,
+      output: `
+        const todoQueries = {
+          key: (a, b, c, d, e) => ({
+            queryKey: ["entity", a, [b], { c }, 1, true, d, e],
+            queryFn: () => api.getEntity(a, b, c, d, e)
+          })
+        }
+      `,
+      errors: [{ messageId: 'missingDeps', data: { deps: 'd, e' } }],
+    },
+    {
+      name: 'should fail when dep does not exist while having a complex queryKey #2',
+      code: `
+        const todoQueries = {
+          key: (dep1, dep2, dep3, dep4, dep5, dep6, dep7, dep8) => ({
+            queryKey: ['foo', {dep1, dep2: dep2, bar: dep3, baz: [dep4, dep5]}, [dep6, dep7]],
+            queryFn: () => api.getEntity(dep1, dep2, dep3, dep4, dep5, dep6, dep7, dep8),
+          }),
+        };
+      `,
+      output: `
+        const todoQueries = {
+          key: (dep1, dep2, dep3, dep4, dep5, dep6, dep7, dep8) => ({
+            queryKey: ['foo', {dep1, dep2: dep2, bar: dep3, baz: [dep4, dep5]}, [dep6, dep7], dep8],
+            queryFn: () => api.getEntity(dep1, dep2, dep3, dep4, dep5, dep6, dep7, dep8),
+          }),
+        };
+      `,
+      errors: [{ messageId: 'missingDeps', data: { deps: 'dep8' } }],
+    },
+    {
+      name: 'should fail when two deps that depend on each other are missing',
+      code: `
+          function Component({ map, key }) {
+            useQuery({ queryKey: ["key"], queryFn: () => api.get(map[key]) });
+          }
+        `,
+      output: `
+          function Component({ map, key }) {
+            useQuery({ queryKey: ["key", map[key]], queryFn: () => api.get(map[key]) });
+          }
+        `,
+      errors: [{ messageId: 'missingDeps', data: { deps: 'map[key]' } }],
+    },
+  ],
```

Comment:
another edge-case that just came to my mind that is actually legit and might yield false positives is:

```
useQuery({
  queryKey: ['foo', dep1],
  queryFn: ({ queryKey: [,dep] }) => api.get(dep)
```

because the `queryKey` is passed to the `queryFn` inside the `queryFunctionContext` - the object that the `queryFn` receives as input.

Reply:
I wish I knew this approach earlier. It would've make typescript happier... I believe I ignore all of the "local dependencies", but I'll have to check that once I get back to it.

Reply:
added a test case. It seems to be working as expected (not adding dep1 to the deps)

Reply:
nice 🚀